source/code/source/Events/Manager/EventPipeline.cpp

```diff
@@ -0,0 +1,304 @@
+{
+}
+
+WriteEventsProvider::~WriteEventsProvider()
+{
+    // Invoking m_terminationCallback here rather than in XAsync callback because the provider may be destroyed prior to
+    // the XAsync callback being invoked. XAsync framework aggressively cleans up the provider as soon it can, and because the operation
+    // has no result payload, cleanup happens immediately after DoWork (prior to the completion callback being invoked in some cases)
+
+    if (m_terminationCallback)
+    {
+        m_terminationCallback();
+    }
+}
+
+HRESULT WriteEventsProvider::Terminate(TerminationCallback callback)
+{
+    if (m_terminated)
+    {
+        // Calling Terminate twice is not allowed
+        return E_PLAYFAB_EVENTMANAGERTERMINATED;
+    }
+
+    m_terminationCallback = std::move(callback);
+    XAsyncCancel(&m_asyncBlock);
+    return S_OK;
+}
+
+HRESULT WriteEventsProvider::DoWork(TaskQueue&& queue)
+{
+    // XAsync should never call XAsyncOp::DoWork after XAsyncOp::Cancel
+    assert(!m_terminated);
+
+    SharedPtr<EventPipeline> pipeline = m_pipeline.lock();
+    if (!pipeline)
+    {
+        // If the pipeline has been destroyed, cancel the task and return E_PENDING. WriteEventsProvider::Cancel 
+        // ensures pending events & batches in flight complete before completing the XAsync operation
+        XAsyncCancel(&m_asyncBlock);
+        return E_PENDING;
+    }
+
+    m_batch.reserve(pipeline->settings.maxItemsInBatch);
+
+    if (m_batchesInFlight >= pipeline->settings.maxBatchesInFlight)
+    {
+        // do not take new events from m_buffer if batches currently in flight are at the maximum allowed number
+        // and are not sent out (or received an error) yet
+        RETURN_IF_FAILED(Schedule(pipeline->settings.waitTimeInMs));
+        return E_PENDING;
+    }
+
+    auto getFromBufferResult = pipeline->m_buffer.TryTake();
+    if (Succeeded(getFromBufferResult))
+    {
+        // Successfully took event from buffer. Add it to current batch
+        m_batch.emplace_back(getFromBufferResult.ExtractPayload());
+
+        // If it's the first event in an incomplete batch then set the batch creation time
+        if (m_batch.size() == 1)
+        {
+            m_batchCreationTime = Clock::now();
+        }
+
+        // Send batch if its full
+        if (m_batch.size() >= pipeline->settings.maxItemsInBatch)
+        {
+            SendBatch(std::move(m_batch), queue);
+        }
+
+        RETURN_IF_FAILED(Schedule(0));
+        return E_PENDING;
+    }
+
+    if (!m_batch.empty())
+    {
+        // check if the batch wait time expired
+        std::chrono::seconds batchAge = std::chrono::duration_cast<std::chrono::seconds>(Clock::now() - m_batchCreationTime);
+        if (batchAge.count() >= (int32_t)pipeline->settings.maxBatchWaitTimeInSeconds)
+        {
+            // batch wait time elapsed, send incomplete batch
+            SendBatch(std::move(m_batch), queue);
+            RETURN_IF_FAILED(Schedule(0));

```

Comment:
>RETURN_IF_FAILED(Schedule(0)); [](http://example.com/codeflow?start=12&length=30)

maybe should wait waitTimeInMs since there were no new events in this case